### PR TITLE
Fuse MatMulInteger and scale followed

### PR DESCRIPTION
**Pull request by yufenglee**
**Description**: Describe your changes.
Fuse MatMulInteger + scale + bias followed
Add bias for DynamicQuantizeMatMul

### CODE REVIEW

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 1]
```diff
@@ -1705,6 +1710,64 @@ Matrix product that behaves like numpy.matmul: https://docs.scipy.org/doc/numpy-
...
+      .Input(6,
+             "bias",
+             "1D input tensor, whose dimension is same as B's last dimension",
+             "T3",
+             OpSchema::Optional)
+      .Output(0, "Y", "Matrix multiply results from A * B", "T3")
+      .TypeConstraint(
+          "T1",
+          {"tensor(int8)", "tensor(uint8)"},
+          "Constrain input B data type to 8-bit integer tensor.")
```
Comment (tracysh) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1757]: This is input A, right? Should the type constraints use the same "input 'A'" (single quoted) style as above?

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 2]
```diff
@@ -1705,6 +1710,64 @@ Matrix product that behaves like numpy.matmul: https://docs.scipy.org/doc/numpy-
...
+          OpSchema::Optional)
+      .Input(
+          5,
+          "b_zero_point",
+          "Zero point tensor for input 'B'. It's optional and default value is 0.  It could be a scalar or a 1-D tensor, "
+          "which means a per-tensor or per-column quantization. If it's a 1-D tensor, its number "
+          "of elements should be equal to the number of columns of input 'B'.",
+          "T2",
+          OpSchema::Optional)
+      .Input(6,
```
Comment (tracysh) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1748]: run clang-format? the input seems to be indented oddly. should the 6 be on the next line to match the style from your other ops?

File: onnxruntime/contrib_ops/cpu/quantization/matmul_integer_to_float.cc [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+template <typename T1, typename T2>
+Status MatMulIntegerToFloat<T1, T2>::Compute(OpKernelContext* ctx) const {
+  const Tensor* a = ctx->Input<Tensor>(0);
+  const Tensor* b = ctx->Input<Tensor>(1);
+  ORT_ENFORCE(a != nullptr && b != nullptr);
+
+  const Tensor* a_scale_tensor = ctx->Input<Tensor>(2);
+  ORT_ENFORCE(IsScalarOr1ElementVector(a_scale_tensor),
+              "MatmulInteger : input A scale must be a scalar or 1D tensor of size 1. Per-Channel is not supported yet.");
```
Comment (tracysh) [inline onnxruntime/contrib_ops/cpu/quantization/matmul_integer_to_float.cc:39]: fixup operator name?

File: onnxruntime/contrib_ops/cpu/quantization/matmul_integer_to_float.cc [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+REGISTER_MATMUL_INTEGER_TO_FLOAT(int8_t)
+REGISTER_MATMUL_INTEGER_TO_FLOAT(uint8_t)
+
+template <typename T1, typename T2>
+Status MatMulIntegerToFloat<T1, T2>::Compute(OpKernelContext* ctx) const {
+  const Tensor* a = ctx->Input<Tensor>(0);
+  const Tensor* b = ctx->Input<Tensor>(1);
+  ORT_ENFORCE(a != nullptr && b != nullptr);
+
```
Comment (skottmckay) [inline onnxruntime/contrib_ops/cpu/quantization/matmul_integer_to_float.cc:36]: nit: If the schema says these are required inputs the onnx checker should have validated that already

File: onnxruntime/core/optimizer/dynamic_quantize_matmul_fusion.cc [hunk 5]
```diff
@@ -12,6 +12,27 @@ using namespace ONNX_NAMESPACE;
 using namespace ::onnxruntime::common;
 namespace onnxruntime {
 
+inline int64_t GetLastDimOfTensorShapeProto(const TensorShapeProto& shape) {
+  return shape.dim(shape.dim_size() - 1).dim_value();
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/dynamic_quantize_matmul_fusion.cc:16]: Can these local helpers be 'static'?

File: onnxruntime/core/optimizer/dynamic_quantize_matmul_fusion.cc [hunk 6]
```diff
@@ -12,6 +12,27 @@ using namespace ONNX_NAMESPACE;
...
+}
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/dynamic_quantize_matmul_fusion.cc:17]: Any chance you would see an empty shape (scalar) or something that doesn't have dim_value set here?

File: onnxruntime/core/optimizer/dynamic_quantize_matmul_fusion.cc [hunk 7]
```diff
@@ -94,34 +138,87 @@ Status DynamicQuantizeMatMulFusion::ApplyImpl(Graph& graph, bool& modified, int
...
     }
 
-    std::vector<NodeArg*> input_defs{dql_node_left.MutableInputDefs()[0],
-                                     matmulinteger_node.MutableInputDefs()[1],
-                                     mul_node_right.MutableInputDefs()[1]};
+    // Find bias node
+    Node* add_node = nullptr;
+    // const Node* add_node = FindBiasNode(graph, mul_node, ;
+    if (optimizer_utils::CheckOutputEdges(graph, mul_node, 1)) {
+      const Node* tmp_add_node = graph_utils::FirstChildByType(mul_node, "Add");
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/dynamic_quantize_matmul_fusion.cc:144]: If this fails should there be a 'continue'?

Comment (yufenglee) [reply onnxruntime/core/optimizer/dynamic_quantize_matmul_fusion.cc:144]: yes, if it failes, then the bias will be null. 

---
In reply to: [448807810](https://github.com/microsoft/onnxruntime/pull/4350#discussion_r448807810) [](ancestors = 448807810)

**Pull request merged**